tg-x/network-architecture.md

```diff
@@ -0,0 +1,234 @@
+---
+title: Anoma Network Architecture
+category: distributed-systems
+publish_date: 2024-07-07
+image: media/network-architecture.jpg
+excerpt: A brief introduction to the network architecture of Anoma.
+---
+
+## Introduction
+
+Anoma is researching and developing a set of network protocols
+that facilitate message passing in a distributed system.
+
+The network protocol stack is designed as a standalone component
+generically useful for building distributed systems.
+
+This article serves as a brief introduction and high-level overview of the network architecture.
+
+## Node architecture
+
+The system is based on the *actor model*,
+where each node runs a set of processes, or actors.
+We call these *engines* in our design.
+Engines have state and communicate with each other via asynchronous message passing:.
+An engine may send and respond to messages,
+react to timer events, and spawn other engines.
+
+Each node has a cryptographic identity that uniquely identifies it in the network.
+A node consists of a collection of running engines.
+Local engines communicate with each other directly,
+while they reach remote nodes via the Router and Transport engines.
+
+The Transport engine is responsible for maintaining authenticated and encrypted network connections between nodes
+via a modular transport system that supports various transport protocols with different reliability, latency, and privacy guarantees.
+Transport protocols initially supported include QUIC and TLS,
+later this can be extended with onion-routed, mix, mesh, and delay-tolerant network protocols.
+
+The Router engine routes incoming and outgoing messages from and to local engines,
+based on the destination address of each message, which consists of a node and engine identity.
+
+<figure>
+
+```mermaid
+flowchart LR
+  T((Transport))
+  R(((Router)))
+  X((Engine_X))
+  Y((Engine_Y))
+
+  X & Y --- R --- T
+```
+
+<figcaption>A node with its engines.</figcaption>
+</figure>
+
+## Communication patterns
+
+The network supports unicast, multicast, and anycast communication patterns.
+
+- *Unicast*: direct messages between two engines.
+- *Multicast*: few-to-many messages sent to a publish-subscribe topic and delivered to all subscribers.
+- *Anycast*: messages sent to any known member of a domain.
+
+<figure>
+
+```mermaid
```

Comment:
in this figure, 
ideally we would have two types of edges
- links/connections/direct acquaintances
- message hops

sth. like this 
![image](https://github.com/anoma/anoma-blog/assets/13623940/97dc1930-5a48-45bd-a438-99fd3ae20e1c)

https://excalidraw.com/#json=qcijAZi3QPhjvbSiTntcl,yxzrxo8AKJKJ-E9gJS4_3w